File: 7.Reverse Integer/ReverseInteger.go
/***************************
 **  Runtime: 4ms         **
 **  Memory Usage: 2.2 MB **
 ***************************/

import "math"

func reverse(x int) int {
	var output int = 0 // assign reversed x in output, and declare output as int32 to avoid buffer overflow
	var remainder int = 0
	for x != 0 {
		remainder = x % 10
		x = x / 10
		output = output * 10 + remainder
		// constraint: -2^31 <= output <= 2^31 - 1
		if output > math.MaxInt32 || output < -math.MaxInt32-1 {
			return 0
		}
	}

	return output
}

File: 7.Reverse Integer/ReverseInteger.c
/****************************
 **  Runtime: 0ms          **
 **  Memory Usage: 5.6 MB  **
 ****************************/

int reverse(int x){
    // save reversed x in output
    // avoid buffer overflow: declare output as long int
    int remainder = 0;
    long int output = 0;
    while (x != 0) {
        remainder = x % 10;
        x  = x / 10;
        output = output * 10 + remainder;
    }
    
    // constraint: -2^31 <= x <= 2^31 - 1
    if (output < INT_MIN || output > INT_MAX) return 0;
    
    return (int)output;
}
